src/coloring.rs

```diff
@@ -59,3 +67,209 @@ pub fn graph_greedy_color(py: Python, graph: &graph::PyGraph) -> PyResult<PyObje
+                        cur_color = self.flip_color(c, d, cur_color);
+                        break;
+                    }
+                }
+            }
+        }
+        path
+    }
+
+    fn check_coloring(&self) -> bool {
+        for edge in self.graph.edge_references() {
+            match self.colors.get(&edge.id()) {
+                Some(_color) => (),
+                None => {
+                    println!("Problem edge {:?} has no color assigned", edge);
+                    return false;
+                }
+            }
+        }
+
+        let mut max_color = 0;
+        for node in self.graph.node_indices() {
+            let mut used_colors: HashSet<usize> = HashSet::new();
+            let mut num_edges = 0;
+            for edge in self.graph.edges(node) {
+                num_edges += 1;
+                match self.colors.get(&edge.id()) {
+                    Some(color) => {
+                        used_colors.insert(*color);
+                        if max_color < *color {
+                            max_color = *color;
+                        }
+                    }
+                    None => {
+                        println!("Problem: edge {:?} has no color assigned", edge);
+                        return false;
+                    }
+                }
+            }
+            if used_colors.len() < num_edges {
+                println!("Problem: node {:?} does not have enough colors", node);
+                return false;
+            }
+        }
+
+        println!("Coloring is OK, max_color = {}", max_color);
+        true
+    }
+
+    pub fn run_algorithm(&mut self) -> &DictMap<EdgeIndex, usize> {
+        for edge in self.graph.edge_references() {
+            let u: NodeIndex = edge.source();
+            let v: NodeIndex = edge.target();
+            let fan = self.get_maximal_fan(edge.id(), u, v);
+            let c = self.get_free_color(u);
+            let d = self.get_free_color(fan.last().unwrap().1);
+
+            // find cdu-path
+            let cdu_path = self.get_cdu_path(u, d, c);
+
+            // invert colors on cdu-path
+            for (edge_index, color) in cdu_path {
+                let flipped_color = self.flip_color(c, d, color);
+                self.colors.insert(edge_index, flipped_color);
+            }
+
+            // find sub-fan fan[0..w] such that d is free on fan[w]
+            let mut w = 0;
+            for (i, (_, z)) in fan.iter().enumerate() {
+                if self.is_free_color(*z, d) {
+                    w = i;
+                    break;
+                }
+            }
+
+            // rotate fan
+            for i in 1..w + 1 {
+                let next_color = self.colors.get(&(fan[i].0)).unwrap();
+                self.colors.insert(fan[i - 1].0, *next_color);
+            }
+
+            // fill additional color
+            self.colors.insert(fan[w].0, d);
+        }
+
+        self.check_coloring();
```

Comment:
Do you plan to expose this to python users or is it just for testing? If you think python users will want this we can add a flag to the pyfunction and put this behind an if statement. If you think it's only good for testing we can put this behind a `debug_assert!()` (https://doc.rust-lang.org/std/macro.debug_assert.html) or something so it only is run when the package is built in debug mode.

Reply:
This was only used for testing. I have moved this function under `#[cfg(test)]`.